File: 0219-contains-duplicate-ii/0219-contains-duplicate-ii.java
class Solution {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
      Map<Integer, Integer> numIndexMap = new HashMap<>();
        
        for (int i = 0; i < nums.length; i++) {
            int num = nums[i];
            if (numIndexMap.containsKey(num)) {
                if (i - numIndexMap.get(num) <= k) {
                    return true;
                }
            }
            numIndexMap.put(num, i);
        }
        
        return false;
    }
}

File: 0219-contains-duplicate-ii/0219-contains-duplicate-ii.py
class Solution(object):
    def containsNearbyDuplicate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: bool
        """
        
        p1, p2 = 0, 0
        myset = set() 

        for p2 in range(len(nums)):
            if nums[p2] in myset:
                return True
            
            myset.add(nums[p2])
            if abs(p1-p2) >= k:
                myset.remove(nums[p1])
                p1+=1

        return False
